### PR TITLE
jtx::sign needs to use addWithoutSigningFields()

### PR DESCRIPTION
**Pull request by scottschurr**
When jtx is generating a signature it can't include signing fields if they are present.  In typical usage jtx does not have any signing fields in place when the signature was generated, so it wasn't an issue.  But for special debugging cases those fields may be already in position.  This change supports both the standard usage model and special debugging cases.  A unit test is included to demonstrate the failing case.

reviewers: @vinniefalco, @ximinez

### CODE REVIEW

**Comment by vinniefalco** [comment]
Suggested commit message:
`Use addWithoutSigningFields in jtx::sign`
:+1:

**Comment by scottschurr** [comment]
Updated commit message.

File: src/ripple/test/jtx/impl/Env_test.cpp [hunk 1]
```diff
@@ -561,6 +561,21 @@ class Env_test : public beast::unit_test::suite
                 );
     }
 
+    void testSig()
+    {
+        using namespace jtx;
+        Env env(*this);
+
+        env.fund(XRP(10000), "alice");
+        auto const baseFee = env.app().config().FEE_DEFAULT;
```
Comment (ximinez) [inline src/ripple/test/jtx/impl/Env_test.cpp]: The expected / canonical way to get fees out of an `Env` is `auto const baseFee = env.open()->fees().base;`.

File: src/ripple/test/jtx/impl/Env_test.cpp [hunk 2]
```diff
@@ -561,6 +561,21 @@ class Env_test : public beast::unit_test::suite
...
+        std::uint32_t const aliceSeq = env.seq ("alice");
+        
+        Json::Value jsonNoop = env.json (
+            noop ("alice"), fee(baseFee), seq(aliceSeq), sig("alice"));
+        JTx jt = env.jt (jsonNoop);
+        env (jt);
```
Comment (ximinez) [inline src/ripple/test/jtx/impl/Env_test.cpp]: So there's a problem with either the change or the test:
1. `jsonNoop` is a `Json` object that includes a signature.
2. That `Json` object is passed to `env.jt()`, which generates a new `JTx`, which is resigned, ignoring the original signature from `jsonNoop`.
3. That `JTx` is submitted to `env`, which generates _yet another_ `JTx`, which is signed yet again, ignoring both original signatures.

In other words, the test will pass, no matter who signs `jsonNoop`. I added some more test cases at https://github.com/ximinez/rippled/commit/f11f00c85d6261365d5a519d32dace8c50d8f236, which demonstrate what I'm talking about. (They all pass.)

If that's intentional, great. Maybe cherry-pick those test cases so it's clearer to future readers. If not...

File: src/ripple/test/jtx/impl/Env_test.cpp (hunk 1 above)
Comment (scottschurr) [inline src/ripple/test/jtx/impl/Env_test.cpp]: Good to know.  I'll fix that.

File: src/ripple/test/jtx/impl/Env_test.cpp (hunk 2 above)
Comment (vinniefalco) [inline src/ripple/test/jtx/impl/Env_test.cpp]: `Env_test.cpp` is not supposed to test business logic, its supposed to test the _jtx_ framework itself. Does this belong here?

Comment (scottschurr) [inline src/ripple/test/jtx/impl/Env_test.cpp]: It seemed like a reasonable place to me.  The test validates a jtx-specific change in `src/ripple/test/jtx/impl/utility.cpp`.  You're welcome to suggest a better place.

Comment (ximinez) [inline src/ripple/test/jtx/impl/Env_test.cpp]: I believe so. It's testing whether _jtx_ properly handles an already-signed transaction as input. The need to look at the fee is a side effect of pre-signing.

Comment (scottschurr) [inline src/ripple/test/jtx/impl/Env_test.cpp]: @ximinez, thanks for the review.  You're pointing out that, at the very least, this test case needs comments and possibly a better name.  This particular test case is intended to test whether jtx can re-sign a transaction that has already been signed.  Without the change in utility.cpp this test fails, but it passes with the change.

Without the change in utility.cpp five of your tests fail.  With the change in utility.cpp all of your tests pass.  I haven't looked at your tests in detail, but it seems like passing is the correct behavior.  Am I understanding your tests correctly?

I'll commit a FOLD that includes comments.  You can tell me whether the comments are sufficient.

Comment (vinniefalco) [inline src/ripple/test/jtx/impl/Env_test.cpp]: @scottschurr Sounds like Env_test.cpp is the right place then, you justification is sound

Comment (ximinez) [inline src/ripple/test/jtx/impl/Env_test.cpp]: Short answer: Comments are probably sufficient.

Longer answer: Yes, without the change in utility.cpp, all 5 of my test cases should fail, because they all start with a signed transaction. The reason I wrote them up was because i was exploring the mechanism of _why_ your test case passes with the change, and noticed that your test transaction's signature was being ignored in favor of re-signing. My test cases demonstrate that a bit more (particularly the second, which is signed by "bob"), but those probably *do* cross over into business logic, and thus may be more relevant elsewhere.

**Comment by ximinez** [comment]
:+1:
